### ISSUE TITLE
Health check support for long provisioned resources

### ISSUE THREAD

**Issue by chanwit** [chanwit/tf-controller#44]
Health check for external resources like Redis, RDS or other managed services like Mongo Atlas requires an extra step to perform a TCP connection to ensure that the target service is already up and running.

The Flagger project has some generic way to declare this test step, from which we could learn.

Our resources maybe something like TCP services, HTTP services. Here's a health check spec proposal.

```yaml
apiVersion: infra.contrib.fluxcd.io/v1alpha1
kind: Terraform
spec:
  path: ./terraform
  healthChecks:
  - name: rds
    kind: TCP
    address: ${output.rds_ip_address}:3306
    timeout: 15m
  - name: nginx
    kind: HTTP
    address: https://${output.nginx_ip_address}/ping
    timeout: 15m
```

**Comment by tomhuang12** [comment]
Could you provide a little more details on what you are envisioning?

**Comment by tomhuang12** [comment]
Do you mean by flagger's [metrics analysis](https://docs.flagger.app/usage/metrics)?

**Comment by chanwit** [comment]
This part:
https://docs.flagger.app/usage/webhooks

**Comment by chanwit** [comment]
Maybe we can focus only at network-based health checks first. Something like this.

```yaml
apiVersion: infra.contrib.fluxcd.io/v1alpha1
kind: Terraform
spec:
  path: ./terraform
  healthchecks:
  - name: rds
    kind: TCP
    address: ${output.rds_ip_address}:3306
    timeout: 15m
  - name: nginx
    kind: HTTP
    address: https://${output.nginx_ip_address}/ping
    timeout: 15m
```

**Comment by tomhuang12** [comment]
Thanks for the details. I implemented the basics of the health checks in the draft PR above. I would like to understand more on in which conditions the health checks should be performed.

IMO, they should be performed after `terraform apply`. Should the health check result change Ready state and any conditions?

**Comment by chanwit** [comment]
```
  - name: nginx
    type: httpPost
    url: https://${output.nginx_ip_address}/post
    timeout: 15m
```
We should not include the HTTP Post method there as a POST normally requires a body content. Please goes with HTTP and assume that it's always the GET method to make thing simple.

> In which state/conditions the health checks should be performed? After apply?

Yes, after Apply succeeded.
  * The condition state of HealthCheck is recorded. HealthCheck = unknown. The reason is "HealthChecking".

> Should the health check result change Ready state and any conditions? Add a new healthCheck condition?

Yes to both. 
  * The Ready state become Ready = true, and the reason changed from `AppliedSucceed` to `HealthCheckSucceed`.
  * The condition state of HealthCheck is recorded. HealthCheck = true. The reason can be the same value as above. 

> Does it perform a health check on every loop?

  * No, we could add an if to say that it's checking only after the Apply step is perfomed.

> When the health check succeed, what should it do in the next loop?

  * When health check done, please return the Terraform object with a nil error. This will pass thru the reconcile login and it's going normally with `Result{RetryAfter: the interval period}`.

> When the health check fails, what should it do in the next loop?

  * When checking fails, we'll do it another round with `Result{RetryAfter:  health check interval }` with an error

> What if the user makes a change to the terraform file when the health check is in a failed state?

  * The current logic should make it thru the normal loop. We could add an if condition to fix that things go wrong.

**Comment by phoban01** [comment]
Might want to consider adding a `successThreshold` and  `failureThreshold` to avoid transient network errors triggering health-check failures? 

Generally, I think the application of tcp/http healthchecks could be limited, as it might often be the case that there's no network connectivity between the resources being created and the tf-controller. If outputs could be passed to exec then something like the following would be more practical:
```yaml
exec: "aws s3api head-bucket --bucket {{ outputs.bucket_name }}"
```

**Comment by chanwit** [comment]
Generic health check would be a great idea, @phoban01 
I'll have another issue to track it as a follow up of this one!

**Comment by tomhuang12** [comment]
> When checking fails, we'll do it another round with Result{RetryAfter: health check interval } with an error
Does that mean we need an another `Interval` param for each health check?

**Comment by chanwit** [comment]
Interesting. Could we?

**Comment by tomhuang12** [comment]
To get the output url, can I use the go template format `{{ outputs.url }}` instead of terraform's format?

**Comment by chanwit** [comment]
Of course, please use what you see fit.

Mine was just an example.

**Comment by tomhuang12** [comment]
Are we good to close this?

**Comment by chanwit** [comment]
Fixed by #53 
Thank you @tomhuang12 !!
